File: sdk/samples/07_advanced/c/content_filter.c
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

/**
 * HDDS Sample: Content Filter (C)
 *
 * Demonstrates content-filtered topic concepts.
 * Content filters allow subscribers to receive only data matching
 * SQL-like filter expressions, reducing network and CPU overhead.
 *
 * Usage:
 *     ./content_filter
 *
 * Key concepts:
 * - ContentFilteredTopic creation
 * - SQL filter expressions
 * - Filter parameters
 * - Application-side filtering demo
 *
 * NOTE: CONCEPT DEMO - This sample demonstrates the APPLICATION PATTERN for ContentFilteredTopic.
 * The native ContentFilteredTopic API is not yet exported to the C/C++/Python SDK.
 * This sample uses standard participant/writer/reader API to show the concept.
 */

#include <hdds.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "generated/HelloWorld.h"

#define NUM_SENSORS 10

/* Sensor data with extended fields */
typedef struct {
    uint32_t sensor_id;
    char location[32];
    float temperature;
    float humidity;
} sensor_data_t;

/* Simple filter structure */
typedef struct {
    float temp_threshold;
    float humidity_threshold;
    const char* location_filter;
} content_filter_t;

/* Check if data matches filter (application-side demo) */
int matches_filter(const sensor_data_t* data, const content_filter_t* filter) {
    if (filter->temp_threshold > 0 && data->temperature <= filter->temp_threshold) {
        return 0;
    }
    if (filter->humidity_threshold > 0 && data->humidity <= filter->humidity_threshold) {
        return 0;
    }
    if (filter->location_filter && strlen(filter->location_filter) > 0) {
        if (strcmp(data->location, filter->location_filter) != 0) {
            return 0;
        }
    }
    return 1;
}

void print_filter_info(void) {
    printf("--- Content Filter Overview ---\n\n");
    printf("Content filters use SQL-like WHERE clause syntax:\n\n");
    printf("  Filter Expression          | Description\n");
    printf("  ---------------------------|---------------------------\n");
    printf("  temperature > 25.0         | High temperature readings\n");
    printf("  location = 'Room1'         | Specific location only\n");
    printf("  sensor_id BETWEEN 1 AND 10 | Sensor ID range\n");
    printf("  humidity > %%0              | Parameterized threshold\n");
    printf("  location LIKE 'Building%%'  | Pattern matching\n");
    printf("\n");
    printf("Note: Full content filter implementation via HDDS extensions.\n");
    printf("This sample demonstrates the filtering concept.\n\n");
}

int main(int argc, char* argv[]) {
    (void)argc;
    (void)argv;

    printf("============================================================\n");
    printf("Content Filter Demo\n");
    printf("SQL-like filtering for DDS topics\n");
    printf("============================================================\n\n");
    printf("NOTE: CONCEPT DEMO - Native ContentFilteredTopic API not yet in SDK.\n");
    printf("      Using standard pub/sub API to demonstrate the pattern.\n\n");

    hdds_logging_init(HDDS_LOG_INFO);

    print_filter_info();

    /* Create participant */
    struct HddsParticipant* participant = hdds_participant_create("ContentFilterDemo");
    if (!participant) {
        fprintf(stderr, "Failed to create participant\n");
        return 1;
    }

    printf("[OK] Participant created: %s\n", hdds_participant_name(participant));

    /* Create endpoints */
    struct HddsDataWriter* writer = hdds_writer_create(participant, "SensorData");
    struct HddsDataReader* reader = hdds_reader_create(participant, "SensorData");

    if (!writer || !reader) {
        fprintf(stderr, "Failed to create endpoints\n");
        if (writer) hdds_writer_destroy(writer);
        if (reader) hdds_reader_destroy(reader);
        hdds_participant_destroy(participant);
        return 1;
    }

    printf("[OK] DataWriter and DataReader created for 'SensorData' topic\n\n");

    /* Define filters */
    printf("--- Defining Content Filters ---\n\n");

    content_filter_t high_temp_filter = {
        .temp_threshold = 30.0f,
        .humidity_threshold = 0,
        .location_filter = NULL
    };
    printf("[OK] Filter 1: temperature > 30.0 (high temperature alerts)\n");

    content_filter_t server_room_filter = {
        .temp_threshold = 0,
        .humidity_threshold = 0,
        .location_filter = "ServerRoom"
    };
    printf("[OK] Filter 2: location = 'ServerRoom'\n");

    content_filter_t combined_filter = {
        .temp_threshold = 25.0f,
        .humidity_threshold = 60.0f,
        .location_filter = NULL
    };
    printf("[OK] Filter 3: temperature > 25.0 AND humidity > 60.0\n\n");

    /* Generate and publish sensor data */
    printf("--- Publishing Sensor Data ---\n\n");

    const char* locations[] = {"ServerRoom", "Office1", "Lobby", "DataCenter"};
    sensor_data_t samples[NUM_SENSORS];
    srand((unsigned int)time(NULL));

    for (int i = 0; i < NUM_SENSORS; i++) {
        samples[i].sensor_id = (uint32_t)(i + 1);
        strncpy(samples[i].location, locations[i % 4], sizeof(samples[i].location) - 1);
        samples[i].temperature = 20.0f + (float)(rand() % 200) / 10.0f;  /* 20-40 */
        samples[i].humidity = 40.0f + (float)(rand() % 400) / 10.0f;      /* 40-80 */

        printf("Publishing: sensor=%u, loc=%s, temp=%.1f, hum=%.1f\n",
               samples[i].sensor_id, samples[i].location,
               samples[i].temperature, samples[i].humidity);

        /* Publish via real HDDS (using HelloWorld as carrier) */
        HelloWorld msg = {.id = (int32_t)samples[i].sensor_id};
        snprintf(msg.message, sizeof(msg.message), "%s:%.1f:%.1f",
                 samples[i].location, samples[i].temperature, samples[i].humidity);

        uint8_t buffer[256];
        size_t len = HelloWorld_serialize(&msg, buffer, sizeof(buffer));
        hdds_writer_write(writer, buffer, len);
    }

    /* Show which samples match each filter */
    printf("\n--- Filter Results (Application-Side) ---\n\n");

    printf("High Temperature Filter (temp > 30.0):\n");
    int match_count = 0;
    for (int i = 0; i < NUM_SENSORS; i++) {
        if (matches_filter(&samples[i], &high_temp_filter)) {
            printf("  [MATCH] sensor=%u, temp=%.1f\n",
                   samples[i].sensor_id, samples[i].temperature);
            match_count++;
        }
    }
    if (match_count == 0) printf("  (no matches)\n");

    printf("\nServerRoom Filter (location = 'ServerRoom'):\n");
    match_count = 0;
    for (int i = 0; i < NUM_SENSORS; i++) {
        if (matches_filter(&samples[i], &server_room_filter)) {
            printf("  [MATCH] sensor=%u, loc=%s\n",
                   samples[i].sensor_id, samples[i].location);
            match_count++;
        }
    }
    if (match_count == 0) printf("  (no matches)\n");

    printf("\nEnvironment Alert Filter (temp > 25 AND hum > 60):\n");
    match_count = 0;
    for (int i = 0; i < NUM_SENSORS; i++) {
        if (matches_filter(&samples[i], &combined_filter)) {
            printf("  [MATCH] sensor=%u, temp=%.1f, hum=%.1f\n",
                   samples[i].sensor_id, samples[i].temperature, samples[i].humidity);
            match_count++;
        }
    }
    if (match_count == 0) printf("  (no matches)\n");

    /* Demonstrate dynamic filter update */
    printf("\n--- Dynamic Filter Update ---\n\n");
    printf("Changing high temperature threshold from 30.0 to 35.0...\n");

    high_temp_filter.temp_threshold = 35.0f;
    printf("[OK] Filter updated dynamically\n");

    printf("\nNew matches (temp > 35.0):\n");
    match_count = 0;
    for (int i = 0; i < NUM_SENSORS; i++) {
        if (matches_filter(&samples[i], &high_temp_filter)) {
            printf("  [MATCH] sensor=%u, temp=%.1f\n",
                   samples[i].sensor_id, samples[i].temperature);
            match_count++;
        }
    }
    if (match_count == 0) printf("  (no matches)\n");

    /* Benefits summary */
    printf("\n--- Content Filter Benefits ---\n\n");
    printf("1. Network Efficiency: Filtering at source reduces traffic\n");
    printf("2. CPU Efficiency: Subscriber processes only relevant data\n");
    printf("3. Flexibility: SQL-like expressions for complex filters\n");
    printf("4. Dynamic Updates: Change filters without recreating readers\n");
    printf("5. Parameterization: Use %%0, %%1 for runtime values\n");

    /* Cleanup */
    hdds_reader_destroy(reader);
    hdds_writer_destroy(writer);
    hdds_participant_destroy(participant);

    printf("\n=== Content Filter Demo Complete ===\n");
    return 0;
}

File: sdk/samples/07_advanced/python/content_filter.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
Content Filter Sample - Demonstrates content-filtered topics

Content filters allow subscribers to receive only data matching
SQL-like filter expressions, reducing network and CPU overhead.

Key concepts:
- ContentFilteredTopic creation
- SQL filter expressions
- Filter parameters
- Dynamic filter updates

Note: This sample demonstrates content filtering patterns. The actual
filtering is performed client-side since HDDS topic-level content
filtering requires application-level implementation.

NOTE: CONCEPT DEMO - This sample demonstrates the APPLICATION PATTERN for ContentFilteredTopic.
The native ContentFilteredTopic API is not yet exported to the C/C++/Python SDK.
This sample uses standard participant/writer/reader API to show the concept.
"""

import os
import sys
import random
import time
import struct
from dataclasses import dataclass
from typing import List, Optional

# Add SDK to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', '..', 'python'))

import hdds


@dataclass
class SensorData:
    """Sensor data type"""
    sensor_id: int = 0
    location: str = ""
    temperature: float = 0.0
    humidity: float = 0.0
    timestamp: int = 0

    def serialize(self) -> bytes:
        """Serialize to bytes."""
        location_bytes = self.location.encode('utf-8')
        # Format: sensor_id(i), loc_len(I), location(s), temp(d), humidity(d), timestamp(q)
        return struct.pack(
            f'<iI{len(location_bytes)}sddq',
            self.sensor_id,
            len(location_bytes),
            location_bytes,
            self.temperature,
            self.humidity,
            self.timestamp
        )

    @classmethod
    def deserialize(cls, data: bytes) -> 'SensorData':
        """Deserialize from bytes."""
        sensor_id, loc_len = struct.unpack_from('<iI', data, 0)
        offset = 8
        location = data[offset:offset + loc_len].decode('utf-8')
        offset += loc_len
        temperature, humidity, timestamp = struct.unpack_from('<ddq', data, offset)
        return cls(sensor_id, location, temperature, humidity, timestamp)


class ContentFilter:
    """Content filter for client-side filtering."""

    def __init__(self, expression: str, parameters: List[str]):
        self.expression = expression
        self.parameters = parameters

    def matches(self, data: SensorData) -> bool:
        """Check if data matches the filter expression."""
        # Simple expression parser for demonstration
        expr = self.expression
        for i, param in enumerate(self.parameters):
            expr = expr.replace(f'%{i}', param)

        # Parse simple conditions
        if ' AND ' in expr:
            parts = expr.split(' AND ')
            return all(self._eval_condition(p.strip(), data) for p in parts)
        elif ' OR ' in expr:
            parts = expr.split(' OR ')
            return any(self._eval_condition(p.strip(), data) for p in parts)
        else:
            return self._eval_condition(expr, data)

    def _eval_condition(self, cond: str, data: SensorData) -> bool:
        """Evaluate a single condition."""
        if '>' in cond:
            field, value = cond.split('>')
            field = field.strip()
            value = float(value.strip())
            return getattr(data, field, 0) > value
        elif '<' in cond:
            field, value = cond.split('<')
            field = field.strip()
            value = float(value.strip())
            return getattr(data, field, 0) < value
        elif '=' in cond:
            field, value = cond.split('=')
            field = field.strip()
            value = value.strip().strip("'\"")
            return str(getattr(data, field, '')) == value
        return False


def print_filter_info():
    print("--- Content Filter Overview ---\n")
    print("Content filters use SQL-like WHERE clause syntax:\n")
    print("  Filter Expression          | Description")
    print("  ---------------------------|---------------------------")
    print("  temperature > 25.0         | High temperature readings")
    print("  location = 'Room1'         | Specific location only")
    print("  sensor_id BETWEEN 1 AND 10 | Sensor ID range")
    print("  humidity > %0              | Parameterized threshold")
    print("  location LIKE 'Building%'  | Pattern matching")
    print()


def main():
    print("=== HDDS Content Filter Sample ===\n")
    print("NOTE: CONCEPT DEMO - Native ContentFilteredTopic API not yet in SDK.")
    print("      Using standard pub/sub API to demonstrate the pattern.\n")

    print_filter_info()

    # Initialize logging
    hdds.logging.init(hdds.LogLevel.INFO)

    # Create participant
    participant = hdds.Participant("ContentFilterDemo")
    print("[OK] Participant created")

    # Create writer with reliable QoS
    writer = participant.create_writer("SensorData", qos=hdds.QoS.reliable())
    print("[OK] DataWriter created (publishes all sensor data)")

    # Create reader
    reader = participant.create_reader("SensorData", qos=hdds.QoS.reliable())
    print("[OK] DataReader created\n")

    # Define content filters (client-side filtering)
    print("--- Creating Content Filters ---\n")

    # Filter 1: High temperature
    high_temp_filter = ContentFilter("temperature > %0", ["30.0"])
    print("[OK] Filter 1: temperature > 30.0 (high temperature alerts)")

    # Filter 2: Specific location
    server_room_filter = ContentFilter("location = %0", ["ServerRoom"])
    print("[OK] Filter 2: location = 'ServerRoom'")

    # Filter 3: Combined filter
    alert_filter = ContentFilter("temperature > %0 AND humidity > %1", ["25.0", "60.0"])
    print("[OK] Filter 3: temperature > 25.0 AND humidity > 60.0\n")

    # Generate and publish sensor data
    print("--- Publishing Sensor Data ---\n")

    locations = ["ServerRoom", "Office1", "Lobby", "DataCenter"]
    samples: List[SensorData] = []

    for i in range(10):
        data = SensorData(
            sensor_id=i + 1,
            location=locations[i % 4],
            temperature=random.uniform(20.0, 40.0),
            humidity=random.uniform(40.0, 80.0),
            timestamp=int(time.time() * 1000)
        )
        samples.append(data)

        print(f"Publishing: sensor={data.sensor_id}, loc={data.location}, "
              f"temp={data.temperature:.1f}, hum={data.humidity:.1f}")

        writer.write(data.serialize())

    # Allow time for data to be received
    time.sleep(0.1)

    # Show filter results
    print("\n--- Filter Results ---\n")

    print("High Temperature Filter (temp > 30.0):")
    for s in samples:
        if high_temp_filter.matches(s):
            print(f"  [MATCH] sensor={s.sensor_id}, temp={s.temperature:.1f}")

    print("\nServerRoom Filter (location = 'ServerRoom'):")
    for s in samples:
        if server_room_filter.matches(s):
            print(f"  [MATCH] sensor={s.sensor_id}, loc={s.location}")

    print("\nEnvironment Alert Filter (temp > 25 AND hum > 60):")
    for s in samples:
        if alert_filter.matches(s):
            print(f"  [MATCH] sensor={s.sensor_id}, temp={s.temperature:.1f}, hum={s.humidity:.1f}")

    # Dynamic filter update
    print("\n--- Dynamic Filter Update ---\n")
    print("Changing high temperature threshold from 30.0 to 35.0...")

    high_temp_filter.expression = "temperature > %0"
    high_temp_filter.parameters = ["35.0"]
    print("[OK] Filter updated dynamically")

    print("\nNew matches (temp > 35.0):")
    for s in samples:
        if high_temp_filter.matches(s):
            print(f"  [MATCH] sensor={s.sensor_id}, temp={s.temperature:.1f}")

    # Read any available data from the reader
    print("\n--- Reading Data from Reader ---\n")
    count = 0
    while True:
        data = reader.take()
        if data is None:
            break
        sample = SensorData.deserialize(data)
        # Apply filter on received data
        if high_temp_filter.matches(sample):
            print(f"  [FILTERED] sensor={sample.sensor_id}, temp={sample.temperature:.1f}")
        count += 1
    print(f"  Total samples received: {count}")

    # Benefits summary
    print("\n--- Content Filter Benefits ---\n")
    print("1. Network Efficiency: Filtering at source reduces traffic")
    print("2. CPU Efficiency: Subscriber processes only relevant data")
    print("3. Flexibility: SQL-like expressions for complex filters")
    print("4. Dynamic Updates: Change filters without recreating readers")
    print("5. Parameterization: Use %0, %1 for runtime values")

    print("\n=== Sample Complete ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())
